### ISSUE TITLE
Cannot get "name" on App\Entity\Task|null.

### ISSUE THREAD

**Issue by VincentLanglet** [twigstan/twigstan#156]
I'm getting
```
Cannot get "name" on App\Entity\Task|null.
🔖 getAttribute.unknown
🐘 compiled_translationDetail.php:312
🌱 templates/Dashboard/TranslationV2/translationDetail.html.twig:262
🎯 src/Controller/Dashboard/TranslationV2Controller.php:181
🎯 src/Controller/Dashboard/TranslationV2Controller.php:207
🎯 templates/Dashboard/TranslationV2/index.html.twig:224
```

And it feel wrong to me for two reasons:

First, the code is 
```
{% if translationTask is empty %}
     Foo
{% else %}
     <span>{{ translationTask.name }}</span>
{% endif %}
```
so twigstan should be able to know `translationTask` is not null.

Secondly, I'm using PHPStan level 7, so `Foo|null` error are not yet reported on my codebase.
Since Twigstan use the same phpstan config, it should be level 7 and not report such error yet by the getAttribute rule.

I feel like there is a `RuleLevelHelper::findTypeToCheck` you don't use in your rule

**Comment by ruudk** [comment]
I implemented the [GetAttribute](https://github.com/twigstan/twigstan/blob/main/src/PHPStan/GetAttributeCheck.php) functionality only recently, and made the decision to error on null. But you're correct, this should follow PHPStan's level / config. I'll work on this.

**Comment by VincentLanglet** [comment]
Yes, and I dunno if `{% if translationTask is empty %}` is not fully understood by twigstan yet

**Comment by ruudk** [comment]
It is, at least for `{% if translationTask %}`
https://github.com/twigstan/twigstan/blob/main/tests/EndToEnd/GetAttribute/narrowing.twig

**Comment by VincentLanglet** [comment]
> I implemented the [GetAttribute](https://github.com/twigstan/twigstan/blob/main/src/PHPStan/GetAttributeCheck.php?rgh-link-date=2024-12-03T09%3A48%3A42Z) functionality only recently, and made the decision to error on null. But you're correct, this should follow PHPStan's level / config. I'll work on this.

I have a related error which might not worth opening another issue
```
Neither the property "tone" nor one of the methods "tone()", "gettone()", "istone()", "hastone()" or "__call()" exist and have public access in class "Symfony\Component\Form\FormView".
🔖 getAttribute.notFound
🐘 compiled_form.php:145
🌱 templates/Dashboard/LanguagePair/form.html.twig:98
🎯 src/Controller/Dashboard/ProjectLanguagePairController.php:141
```

FormView is an arrayAccess and currently twigstan reports tons of error for code like
```
 {{ form_label(form.tone) }}
```

**Comment by ruudk** [comment]
It is, I created it for your: https://github.com/twigstan/twigstan/issues/159

I didn't bother yet with ArrayAccess in GetAttribute.

**Comment by VincentLanglet** [comment]
> I implemented the [GetAttribute](https://github.com/twigstan/twigstan/blob/main/src/PHPStan/GetAttributeCheck.php?rgh-link-date=2024-12-03T09%3A48%3A42Z) functionality only recently, and made the decision to error on null. But you're correct, this should follow PHPStan's level / config. I'll work on this.

When reading the current implementation, I feel like the GetAttributeCheck has multiple issues:
- It should use `$this->ruleLevelHelper->findTypeToCheck` for the GetAttributeRule but not for the GetAttributeReturnType

- You take only the first constant string but I'm unsure if it's possible to have more
https://github.com/twigstan/twigstan/blob/1d36397e3843da56da879c7f9f0b4171b0a44a9d/src/PHPStan/GetAttributeCheck.php#L69
at least if would be safer to change `if ($constantStringTypes === []) {` to `if (count($constantStringTypes) !== 1) {`

- When there is an error you returns ErrorType (Like in the Task|null) case, but you should still returns the valid computed return type (here is would be the return type of Task->getName()) in order to display more possible error. For instance
```
{% set name = translationTaskPossiblyNull.name %} // You report Cannot get "name" on App\Entity\Task|null.
{% set foo = thisMethodWantsInteger(name) %} // I think you report nothing since name is ErrorType
```

- I think you don't handle UnionType correctly if it's array|object since you work with `! $objectType->isObject()->yes()` and `$objectType->isArray()->yes()`. And I would think there is certainly other place in this class where UnionType are not well handled
https://github.com/twigstan/twigstan/blob/1d36397e3843da56da879c7f9f0b4171b0a44a9d/src/PHPStan/GetAttributeCheck.php#L94

For the GetAttributeReturnType, it may be better to
- write a getAttributeOnSingleType method (which would be similar to the one you wrote)
- write
```
$type= $scope->getType($arguments['object']);
if ($type instanceof UnionType) {
    $types = [];
    foreach ($type->getTypes() as $singleType) {
        $result = $this->getAttributeOnSingleType($singleType);
        if (!$result[0] instanceof ErrorType) {
            $types[] = $result[0];
        }
    }
    return TypeCombinator::union($types);
} else {
     return $this->getAttributeOnSingleType($type)[0];
}
```

For the GetAttributeRule, it would be
```
$type = $this->ruleLevelHelper->findTypeToCheck(
     $scope,
     $arguments['object']
     '',
     // I dunno what to use as fourth param
);
if ($type instanceof UnionType) {
    // Not sure how to merge the errors
} else {
     return $this->getAttributeOnSingleType($type)[1];
}
```
